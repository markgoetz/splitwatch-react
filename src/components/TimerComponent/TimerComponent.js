import React from 'react';
import renderTime from '../../lib/renderTime';
import TimerContext from '../../lib/TimerContext';

import styles from './TimerComponent.module.css';

export default class TimerComponent extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            elapsedTime: 0
        };
        this.timeout = -1;
    }

    componentDidMount() {
        this.timeout = window.requestAnimationFrame(this.updateTime);
    }

    updateTime = () => {
        const context = this.context;
        this.setState({
            elapsedTime : (context.isPlaying) ?
                Date.now() - context.startTime + context.baseTime :
                context.baseTime
        });
        window.requestAnimationFrame(this.updateTime);
    }

    render() {
        return <div className={styles.timer}>
            { renderTime(this.state.elapsedTime, 1) }
        </div>;
    }
}

TimerComponent.contextType = TimerContext;

// export default function TimerComponent() {
//     let [elapsedTime, setElapsedTime] = useState(0);
//     let timerContext = useContext(TimerContext);

//     useEffect(() => {
//         let timeout;

//         function updateTime() {
//             setElapsedTime((timerContext.isPlaying) ?
//                     Date.now() - timerContext.startTime + timerContext.baseTime :
//                     timerContext.baseTime);

//             if (!timerContext.isPlaying) {
//                 window.clearInterval(timeout);
//             }

//             return () => {
//                 window.clearInterval(timeout);
//             }
//         }
//         timeout = window.setInterval(updateTime, 100);
//     });

//     return (
//         <div className={styles.timer}>
//             { renderTime(elapsedTime, 1) }
//         </div>
//     );
// }
