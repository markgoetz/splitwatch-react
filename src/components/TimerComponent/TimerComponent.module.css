.timer {
    font-family: var(--font-monospace);
    font-size: 52px;
    font-weight: bold;
}
