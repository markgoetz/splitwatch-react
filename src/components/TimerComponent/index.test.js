import React from 'react';
import TimerComponent from './TimerComponent';
import { shallow } from 'preact-render-spy';

const testNow = new Date('2018-01-01');
global.requestAnimationFrame = jest.fn();
let component;

beforeEach(() => {
  component = new TimerComponent({ startTime: 3000, baseTime: 1000, isPlaying: false });
  global.requestAnimationFrame.mockReset();
  require('jest-mock-now')(testNow);
});

describe('TimerComponent', () => {

  test('Renders correctly on play', () => {
    const context = shallow(<TimerComponent isPlaying startTime={3000} baseTime={1000} />);
    expect(context).toMatchSnapshot();
  });

  test('Renders correctly on pause', () => {
    const context = shallow(<TimerComponent startTime={3000} baseTime={1000} />);
    expect(context).toMatchSnapshot();
  });

  test('Calls requestAnimationFrame on mount', () => {
    component.componentDidMount();
    expect(global.requestAnimationFrame).toHaveBeenCalledWith(component.updateTime);
  });

  test('Calls requestAnimationFrame when updateTime is called', () => {
    component.updateTime();
    expect(global.requestAnimationFrame).toHaveBeenCalledWith(component.updateTime);
  });

  test('Updates state when updateTime is called and component is paused', () => {
    component.updateTime();
    expect(component.state).toEqual({ elapsedTime: 1000 });
  });

  test('Updates state when updateTime is called and component is playing', () => {
    component.props.isPlaying = true;
    component.updateTime();
    expect(component.state).toEqual({ elapsedTime: testNow.getTime() - 2000 });
  });
});