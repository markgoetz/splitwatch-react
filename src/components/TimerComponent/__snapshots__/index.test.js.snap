// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerComponent Renders correctly on pause 1`] = `
preact-render-spy (1 nodes)
-------
<div>00:00:00.0</div>

`;

exports[`TimerComponent Renders correctly on play 1`] = `
preact-render-spy (1 nodes)
-------
<div>00:00:00.0</div>

`;
